File: 面试题31-栈的压入、弹出序列.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> mystack;
        int i = 0;

        for(int j = 0; j < popped.size(); j++){
            if(mystack.empty() || mystack.top() != popped[j]){
                while(i < pushed.size() && pushed[i] != popped[j]){
                    mystack.push(pushed[i]);
                    i++;
                }

                if(i == pushed.size())
                    return false;

                i++;
            }
            
            else
                mystack.pop();
        }

        return true;
    }
};

File: 面试题31-栈的压入、弹出序列.py
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        mystack = []
        i = 0

        for j in range(len(popped)):
            if not mystack or mystack[-1] != popped[j]:
                while i < len(pushed) and pushed[i] != popped[j]:
                    mystack.append(pushed[i])
                    i += 1
                
                if i == len(pushed):
                    return False
                
                i += 1

            else:
                mystack.pop()
            
        return True
